Fix part of #2691: Migrate compute affected tests script to Kotlin

BenHenning opened the pull request:
Fix part of #2691

## Explanation
This PR introduces a new version of the compute_affected_tests script that exists in Kotlin instead of Bash.

The main reason for doing this is the new script includes code to fix situations where Bazel outputs malformed query results (e.g. when two targets are on the same line). It's not yet clear whether these issues are caused by Bazel or Bash, but they only happen in CI (i.e. can't be reproduced locally) on occasional PRs (such as #3340). This version of the script builds in resilience to such failures.

There are a number of other benefits that we see moving to Kotlin, as well:
- It's easier to test the script
- The script is normalizing on an existing language well understood in the codebase (Kotlin) vs. using an alternative language (such as Python)
- The script is easier to understand since advanced Bash syntax is hard to read and understand vs. Kotlin which is a much more self-documenting language
- The script establishes the foundation for introducing more Kotlin-based scripts in the future

Beyond that, three other things to note:
- I noticed that the per-file query loop in the original Bash script can be simplified to a single command which seems to be more performant. Beyond that, the script overall seems much faster through Bazel vs. Bash (though CI will be the true test, especially since the binary also needs to be build via Bazel itself).
- The linters have been updated to include the scripts directory as actual code
- I adjusted the output of the ktlint command to include a copy-and-pastable line for fixing errors based on the version of ktlint used to report the errors
- The command runs in a different directory than it did with Bash hence the updated usage arguments. This is actually safer from a sandboxing perspective as it reduces the chance of the script having accidental side effects in the repository.
- The original command may have had a bug in it when an empty set is passed for BUILD files (such as if the script is run with no files changed). This has been addressed in the new script.
- The original script would sometimes hang on CI for unknown reasons. While this could still happen, additional debug lines have been added to the script to help with this & other issues as they crop up. Further, the new script has a timeout if subcommands take longer than 60 seconds, so CI should never indefinitely hang due to the script now.
- The new script is actually testable, and the new tests are creating a real Bazel & Git repository locally in order to simulate different situations.
- The new script is slower since it's no longer in Bash, but it seems fast enough for our purpose (it actually seems to be comparable, but this PR is simpler from a dependency graph perspective than some PRs we expect this to run in).

Long-term, we'll likely want to move all of our scripts over to Kotlin to reduce the total number of languages needed for the repository by 1.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

BenHenning committed:
Migrate compute affected tests script to Kotlin.

No tests for the script yet.

BenHenning committed:
Add BUILD.bazel docstrings.

BenHenning committed:
Enable linters for scripts code. Fix ktlint issues.

BenHenning commented:
@Sparsh1212 are you still hitting the problem we discussed in #3340? It might be nice to do a trial run on your PR, but that might be tricky since the code needs to be copied into the branch and uploaded to GitHub.

Keeping this as a draft until I write tests, though keep in mind that tests are particularly hard for this script since it's highly dependent on a real Bazel environment.

Sparsh1212 commented:
@BenHenning I copied the code from `regex-checks` branch to this branch and initiated a draft PR to see the CI checks are working properly or not.
PR: #3375

Sparsh1212 commented:
The parsing of the added `kt_jvm_test`: `test_pattern_validation_check` is now correctly taking place. Also, the tests are passing as well.

BenHenning commented:
Ah excellent--thanks @Sparsh1212. In that case, this just needs tests & can then go into review.

Sparsh1212 commented:
> Ah excellent--thanks @Sparsh1212. In that case, this just needs tests & can then go into review.

Thanks! for the fix Ben!

BenHenning committed:
Add tests for the compute affected tests script.

Fix some typos & realized bug in the script.

BenHenning commented:
Added tests for the core script. Still need to refactor, added more documentation, and add tests for the test & script utilities.

BenHenning commented:
Also interesting: I changed the merge-base command to use ``develop`` instead of ``origin/develop`` since the test repository doesn't have an origin, but that doesn't work in the CI environment. The CI script probably needs to check out the develop branch so that it exists locally.

BenHenning committed:
Check out develop first so that exists in CI.

See comment in PR for context.

BenHenning commented:
Unfortunately the extra checkout action didn't work. We may need to pass the reference object for merge-base in as a parameter unfortunately.

BenHenning committed:
Reorganize & refactor, and improve tests.

This splits up the test & script into multiple utility packages (for
common & testing utilities), reorganizes the test to be in a proper
javatests/ & java/ arrangement, and adds tests for checking the error
conditions of the test.

BenHenning committed:
Add tests for executeCommand().

This also refactors that function to be part of a class so that its
operating conditions are configurable (namely the process timeout).

BenHenning committed:
Add tests for BazelClient.

BenHenning committed:
Add tests for GitClient.

BenHenning committed:
Add tests for TestBazelWorkspace & documentation.

Also, fix a few small issues in the utility.

BenHenning committed:
Add tests git TestGitRepository.

BenHenning committed:
Merge branch 'develop' into migrate-compute-affected-tests-to-kotlin

BenHenning committed:
Add support for changing relative develop branch.

Also, fix some changed behaviors that broke the affected tests script.

BenHenning committed:
Fix linter error found in CI.

BenHenning commented:
Okay, I think the script is fully ready for review now since it seems to be operating as expected, and it + its constituent components are now fully documented & tested.

BenHenning commented:
PTAL @vinitamurthi @anandwana001 & @Sparsh1212 (FYI that this blocks Sparsh's milestone 1 work since one of his PRs isn't running on CI correctly due to issues with the existing Bash version of the script, and this arranges some of the script directory components which will help normalize patterns in downstream script PRs for his & other GSoC projects).

Sparsh1212 commented on scripts/src/java/org/oppia/android/scripts/testing/TestBazelWorkspace.kt:
```diff
@@ -0,0 +1,231 @@
+    return rootBuildFile.appendingPrintWriter().use { writer ->
+      val depFile = temporaryRootFolder.newFile("$dependencyName.kt")
+      libraryFileMap[libTargetName] = depFile
+      writer.println("kt_jvm_library(")
+      writer.println("    name = \"$libTargetName\",")
+      writer.println("    srcs = [\"${depFile.name}\"],")
+      writer.println(")")
```
or maybe, can we do this?
```kotlin
val ktJvmLibrary = 
"""
kt_jvm_library(
    name = "$libTargetName",
    srcs = ["${depFile.name}"],
)
"""
writer.println(ktJvmLibrary)
```
This might look a bit neat. Ditto elsewhere.

BenHenning replied:
Nice idea--this also let me get rid of the ``appendingPrintWriter`` helper function.

Sparsh1212 commented:
@BenHenning 
The CI checks are failing for some of the `javatests`. PTAL.

anandwana001 reviewed (changes requested):
Everything seems reasonable, just one of the test cases is failing in `BazelClientTest `.

anandwana001 commented on scripts/src/java/org/oppia/android/scripts/ci/BUILD.bazel:
```diff
@@ -0,0 +1,18 @@
+    ],
+    visibility = ["//scripts:oppia_script_binary_visibility"],
+    deps = [
+      "//scripts/src/java/org/oppia/android/scripts/common:bazel_client",
+      "//scripts/src/java/org/oppia/android/scripts/common:git_client",
```
```suggestion
        "//scripts/src/java/org/oppia/android/scripts/common:bazel_client",
        "//scripts/src/java/org/oppia/android/scripts/common:git_client",
```

This will fix the Bazel lint error

BenHenning replied:
Done.

BenHenning commented:
Interesting. The failing test actually may be the same timeout issue we're seeing for certain BUILD file-based PRs for the script in CI. We may have found a deterministic means of reproducing the test in CI (though probably not yet locally unfortunately). I'm going to upload the latest changes to try and see if the test failure repros.

BenHenning committed:
Address reviewer comments.

BenHenning commented:
Looks like this might be consistent: https://github.com/oppia/oppia-android/pull/3374/checks?check_run_id=2976768230.

BenHenning committed:
Remove unnecessary extra checkout from CI.

BenHenning commented:
I was able to isolate the problem. I've added context to https://github.com/bazelbuild/bazel/issues/906. I might need to file a new issue since that one is old, but I'm really interested in seeing whether the isolated repository will make it easier to investigate the issue.

In the meantime, I'll disable the test with a TODO to unblock this & downstream work.

BenHenning committed:
Disable test that can't pass in CI.

See #2691 for context.

BenHenning commented:
Now that the test is ignored, I expect all remaining tests to pass. Will reassign once I confirm the failing suite in question is now passing on CI.

BenHenning commented:
BazelClientTest is now passing: https://github.com/oppia/oppia-android/pull/3374/checks?check_run_id=2976899202. Assigning this back over since the PR should have a fully green CI now.

BenHenning commented:
@Sparsh1212 / @anandwana001 PTAL at my follow-up comments & mark as resolved if you're happy with the changes made.

vinitamurthi reviewed (approved):
LGTM!

Sparsh1212 commented:
LGTM!

anandwana001 reviewed (approved):
LGTM

BenHenning commented:
Thanks all! I might follow up with a workaround to the issue with the query hanging (I have an idea that might mitigate it).
